src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.ConnectionPools.cs

```diff
@@ -0,0 +1,347 @@
+            /// </summary>
+            public static async Task<TResult> RetryRemoteCallAsync<TException, TResult>(
+                Func<Task<TResult>> funcAsync,
+                TimeSpan timeout,
+                CancellationToken cancellationToken) where TException : Exception
+            {
+                const int retry_delayInMS = 50;
+
+                using (var pooledStopwatch = SharedPools.Default<Stopwatch>().GetPooledObject())
+                {
+                    var watch = pooledStopwatch.Object;
+                    watch.Start();
+
+                    while (watch.Elapsed < timeout)
+                    {
+                        cancellationToken.ThrowIfCancellationRequested();
+
+                        try
+                        {
+                            return await funcAsync().ConfigureAwait(false);
+                        }
+                        catch (TException)
+                        {
+                            // throw cancellation token if operation is cancelled
+                            cancellationToken.ThrowIfCancellationRequested();
+                        }
+
+                        // wait for retry_delayInMS before next try
+                        await Task.Delay(retry_delayInMS, cancellationToken).ConfigureAwait(false);
+
+                        ReportTimeout(watch);
+                    }
+                }
+
+                // operation timed out, more than we are willing to wait
+                ShowInfoBar();
+
+                // user didn't ask for cancellation, but we can't fullfill this request. so we
+                // create our own cancellation token and then throw it. this doesn't guarantee
+                // 100% that we won't crash, but this is at least safest way we know until user
+                // restart VS (with info bar)
+                using (var ownCancellationSource = new CancellationTokenSource())
+                {
+                    ownCancellationSource.Cancel();
+                    ownCancellationSource.Token.ThrowIfCancellationRequested();
+                }
+
+                throw ExceptionUtilities.Unreachable;
+            }
+
+            public static async Task<Stream> RequestServiceAsync(
+                HubClient client,
+                string serviceName,
+                HostGroup hostGroup,
+                TimeSpan timeout,
+                CancellationToken cancellationToken = default)
+            {
+                const int max_retry = 10;
+                const int retry_delayInMS = 50;
+
+                Exception lastException = null;
+
+                var descriptor = new ServiceDescriptor(serviceName) { HostGroup = hostGroup };
+
+                // call to get service can fail due to this bug - devdiv#288961 or more.
+                // until root cause is fixed, we decide to have retry rather than fail right away
```

Comment:
i'm confused.  we are calling RetryRemoteCallAsync, but we also have a retry loop here?  In practice, retries within retries are not a good idea.

Reply:
well this is due to wierdness of how service hub works. service hub has its hard coded timeout which throws cancellation token. that is first retry (hiding this behavior of service hub)

..

next retry is another weirdness of service hub. RequesetService can fail time to time for some race issue which throws RemoteInvocationException. the second one is retry of that one. I dont want to merge 2. easier to know what is actual issue (the second retry). and what is expected (the first cancellation exception)

Reply:
we found many issues from the second exception. and service hub team has been fixing those.

Reply:
All this needs to be documented in the code much more clearly.  it's not at all clear why we need the two different retries.  This should all be explained so that anyone else who comes to work in here can understand why this is necessary.  Thanks!

Reply:
sure